### PR TITLE
Display message time.

### PR DESCRIPTION
**Pull request by o01eg**
This PR adds UTC timestamp to server-side chat messages and display them in the local timezone on client.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
Should probably have an option to enable / disable this...

**Comment by o01eg** [comment]
Add options `UI.display-timestamp` and `UI.display-timestamp-format`.

File: UI/ClientUI.cpp [hunk 1]
```diff
@@ -651,6 +656,11 @@ ClientUI::ClientUI() :
 
     // Set the root path for image tags in rich text.
     GG::ImageBlock::SetDefaultImagePath(ArtDir().string());
+
+    // Set facet to format timestamp in chat.
+    boost::posix_time::time_facet* facet = new boost::posix_time::time_facet();
+    facet->format(ClientUI::DisplayTimestampFormat().c_str());
+    m_date_format_sstream.imbue(std::locale(std::locale(""), facet));
```
Comment (geoffthemedio) [inline UI/ClientUI.cpp]: does this call to `imbue` take ownership of `facet`?

Comment (o01eg) [reply UI/ClientUI.cpp]: Yes, it takes:
http://www.cplusplus.com/reference/locale/locale/locale/
> The constructed locale object takes over responsibility for deleting this facet object.

**Comment by o01eg** [comment]
Rebased, allowed UI to show messages without sender.

**Comment by o01eg** [comment]
Rebased. Removed `UI.display-timestamp-format`.

**Comment by geoffthemedio** [comment]
MPLobby chat layout is a bit awkward with the lengthy timestamps before the chat messages...
mplobby_chat_with_timestamps

**Comment by o01eg** [comment]
I increased width of window and chat box.

**Comment by geoffthemedio** [comment]
Can you add an OptionsWnd checkbox to turn the timestamp on / off?

**Comment by o01eg** [comment]
Added in UI section.

File: UI/OptionsWnd.cpp [hunk 2]
```diff
@@ -522,6 +522,7 @@ void OptionsWnd::CompleteConstruction() {
     BoolOption(current_page, 0, "UI.window-quickclose",          UserString("OPTIONS_QUICK_CLOSE_WNDS"));
     BoolOption(current_page, 0, "UI.sidepanel-planet-shown",     UserString("OPTIONS_SHOW_SIDEPANEL_PLANETS"));
     BoolOption(current_page, 0, "UI.auto-reposition-windows",    UserString("OPTIONS_AUTO_REPOSITION_WINDOWS"));
+    BoolOption(current_page, 0, "UI.display-timestamp",          UserString("OPTIONS_DB_DISPLAY_TIMESTAMP"));
```
Comment (geoffthemedio) [inline UI/OptionsWnd.cpp]: The options DB entry is too verbose for the option label. It works as the tooltip, but the options UI label should be short, like just "Chat Timestamps". Have a separate stringtable entry in the game options section and use that for the extra label...

https://github.com/freeorion/freeorion/blob/master/default/stringtables/en.txt#L2325

Comment (o01eg) [reply UI/OptionsWnd.cpp]: Fixed.

File: default/stringtables/en.txt [hunk 3]
```diff
@@ -2340,6 +2343,9 @@ Show side panel planets
 OPTIONS_AUTO_REPOSITION_WINDOWS
 Automatically reposition windows
 
+OPTIONS_DISPLAY_TIMESTAMP
+Show chat timestamp.
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: no `.` (as with similar strings)

Comment (o01eg) [reply default/stringtables/en.txt]: Removed.

**Pull request merged**